validator/attester/service.go

```diff
@@ -22,49 +26,78 @@ type Attester struct {
 	cancel         context.CancelFunc
 	assigner       assignmentAnnouncer
 	assignmentChan chan bool
+	p2p            shared.P2P
+	blockBuf       chan p2p.Message
 }
 
 // Config options for an attester service.
 type Config struct {
-	AssignmentBuf int
-	Assigner      assignmentAnnouncer
+	AssignmentBuf   int
+	Assigner        assignmentAnnouncer
+	BlockBufferSize int
 }
 
 // NewAttester creates a new attester instance.
-func NewAttester(ctx context.Context, cfg *Config) *Attester {
+func NewAttester(ctx context.Context, cfg *Config, validatorP2P shared.P2P) *Attester {
 	ctx, cancel := context.WithCancel(ctx)
 	return &Attester{
 		ctx:            ctx,
 		cancel:         cancel,
 		assigner:       cfg.Assigner,
 		assignmentChan: make(chan bool, cfg.AssignmentBuf),
+		p2p:            validatorP2P,
+		blockBuf:       make(chan p2p.Message, cfg.BlockBufferSize),
 	}
 }
 
 // Start the main routine for an attester.
-func (p *Attester) Start() {
+func (a *Attester) Start() {
 	log.Info("Starting service")
-	go p.run(p.ctx.Done())
+	go a.run(a.ctx.Done())
 }
 
 // Stop the main loop.
-func (p *Attester) Stop() error {
-	defer p.cancel()
+func (a *Attester) Stop() error {
+	defer a.cancel()
 	log.Info("Stopping service")
 	return nil
 }
 
+// CreateAttestation will cause the attester to attest to the provided beacon block.
+func (a *Attester) CreateAttestation(block *pbp2p.BeaconBlock) *pbp2p.AttestationRecord {
+	attestation := &pbp2p.AttestationRecord{}
+	attestation.Slot = block.GetSlotNumber()
+
+	// TODO(#487): Attesters responsibilities will be handled by this PR(#487).
+	attestation.AttesterBitfield = []byte{}
+
+	return attestation
+}
+
 // run the main event loop that listens for an attester assignment.
-func (p *Attester) run(done <-chan struct{}) {
-	sub := p.assigner.AttesterAssignmentFeed().Subscribe(p.assignmentChan)
+func (a *Attester) run(done <-chan struct{}) {
+	sub := a.assigner.AttesterAssignmentFeed().Subscribe(a.assignmentChan)
+	blocksub := a.p2p.Subscribe(&shardingp2p.BlockBroadcast{}, a.blockBuf)
 	defer sub.Unsubscribe()
+	defer blocksub.Unsubscribe()
+
 	for {
 		select {
 		case <-done:
 			log.Debug("Attester context closed, exiting goroutine")
 			return
-		case <-p.assignmentChan:
+		case <-a.assignmentChan:
 			log.Info("Performing attester responsibility")
+		case msg := <-a.blockBuf:
+			data, ok := msg.Data.(*shardingp2p.BlockBroadcast)
+			if !ok {
+				log.Error("Received malformed attestation p2p message")
+				continue
+			}
+			attestation := a.CreateAttestation(data.GetBeaconBlock())
+			a.p2p.Broadcast(attestation)
```

Comment:
This block is not correct. Can you take out this block and refer to #487?

The correct work flow should be:
1. Attester wait for its slot and gets HEAD from beacon node
2. Attester signs HEAD
3. Attester relays HEAD back to beacon node, pass it to rpc service to block chain service to p2p service
4. p2p service will broadcast the attestation

I don't think we should broadcast attestation from client level

Reply:
hmm , so what does the proposer actually do then ? 
Why does the attester pass it back to the beacon node then , only for it to be broadcasted again by the beacon node ?
What is the purpose of the beacon node in the above scenario , I was under the impression that attestations and signatures were all aggregated by both attesters and proposers , with the only difference being that the proposer sends the block back to the beacon node through rpc .

Reply:
I think attester should also send the attestation back to the beacon node for book keeping. Unless attester plans to do it himself, but that'd be duplicated effort for attester and proposer. 

Ideally, beacon node should be managing the attestations (ex: DB look up, save old & new attestation in DB, fix oblique parent hashes and find slashing conditions)

Reply:
Alright let me try to clear up what the flow of our clients will be like : 
Proposer: 
- In charge of block production, will aggregate all the signature from attesters , and then finally send the block through grpc back to the beacon - node 
Attester:
- Will attest to a potential beacon block, adding in shard id, shard blockhash ,etc and adding the attester bitfield
The issue is how will the the flow would go like, if the attester actually sends the attestation back to the beacon - node to do all this other stuff , ex: fix oblique parent hashes , find slashing conditions, etc and then broadcast it to the proposer, then what is actually the proposer's role in all this then ?
All the work is being done by the beacon-node and attester, the proposer then just aggregates the signatures ? Wouldnt that also be possible to be done by the beacon-node by itself then ?
In the client flow we are describing above, I dont see what role the proposer will have, 
We would also not need to have p2p messaging by validators, the beacon node could simply be connected to the attester clients through rpc.

Reply:
Not just proposer, attester broadcasts attestations to the rest of the validators. It's every validator's job to aggregate attestations and keep it among themselves. Only proposer gets to propose the block.

Proposer does more in this context. It needs to sign off on its proposed block. A beacon node can not process an incoming beacon block if proposer's signature is not included.